server/catgenome/src/main/java/com/epam/catgenome/manager/reference/MotifSearchManager.java

```diff
@@ -110,36 +116,72 @@ private void verifyMotifSearchRequest(final MotifSearchRequest motifSearchReques
 
     private MotifSearchResult searchRegionMotifs(final MotifSearchRequest request) {
         final Chromosome chromosome = loadChrById(request.getReferenceId(), request.getChromosomeId());
+        final byte[] sequence;
+        try {
+            sequence= referenceManager.getSequenceByteArray(request.getStartPosition(),
+                            request.getEndPosition(), request.getReferenceId(), chromosome.getName());
+        } catch (IOException e) {
+            throw new IllegalStateException("Unable to read reference data by ID: " + request.getReferenceId(), e);
+        }
+        final List<Motif> searchResult =
+                MotifSearcher.search(sequence, request.getMotif(), request.getStrand(),
+                        chromosome.getName(), request.getStartPosition());
+        final int lastStart = searchResult.size() == 0
+                ? request.getStartPosition()
+                : searchResult.get(searchResult.size() - 1).getStart();
         return MotifSearchResult.builder()
-                .result(
-                    fillMotifList(
-                        chromosome, request.getStartPosition(),
-                        request.getEndPosition(), 0,
-                            request.getStrand()))
+                .result(searchResult)
                 .chromosomeId(request.getChromosomeId())
-                .pageSize(request.getPageSize())
-                .position(request.getEndPosition())
+                .pageSize(searchResult.size())
+                .position(lastStart)
                 .build();
     }
 
     private MotifSearchResult searchChromosomeMotifs(final MotifSearchRequest request) {
         final Chromosome chromosome = loadChrById(request.getReferenceId(), request.getChromosomeId());
-        int start = request.getStartPosition() == null
+        final int start = request.getStartPosition() == null
                 ? 0
                 : request.getStartPosition();
-        int end = request.getEndPosition() == null
+        final int end = request.getEndPosition() == null
                 ? chromosome.getSize()
                 : request.getEndPosition();
 
+        final int bufferSize = Math.min(BUFFER_SIZE, end - start);
+        final int overlap = bufferSize < BUFFER_SIZE ? 0 : OVERLAP;
+        Assert.isTrue(overlap < bufferSize, "overlap must be lower then buffer size!");
+
+        final Set<Motif> result = new LinkedHashSet<>();
+        int currentStart = start;
+        int currentEnd = bufferSize + start;
+
+        while (result.size() < request.getPageSize() && currentEnd <= end ) {
```

Comment:
what if pageSize is not provided? we need to use default value, lets use `100` as default